### PR TITLE
Fixes for latest Garry's Mod update (2023)

### PR DESCRIPTION
**Pull request by Risusama**
This Pull Request aims to fix compatibility with 2023.01 patch of Garry's Mod. (#232)
**Not ready for merge just yet, more changes are needed.**

### CODE REVIEW

**Comment by Risusama** [comment]
I'll look through it all tomorrow again just to be sure I didn't miss anything.
So far there's two occurences that I simply don't know how to fix.

Then there's also console spam like below:
`
projectile_ghoulfleshpuke[300]: Changing collision rules within a callback is likely to cause crashes!
projectile_poisonpuke[301]: Changing collision rules within a callback is likely to cause crashes!
`

**Comment by Risusama** [comment]
So the clientside errors seem to be caused by "CreateWeaponOfQuality" trying to create weapons with qualities like "_q1_q2" and "_q1_q2_q3".

I don't know how to fix that, and honestly don't have the time to learn that much at the moment. :D

**Comment by Alex9914** [comment]
you can add `wept.ClassName = newclass` at ln138 to fix the sh_weaponquality issues (Credit to O Que#7479 for finding and fixing it)

**Comment by Alex9914** [comment]
add wept.ClassName = newclass to ln138 of sh_weaponquaiity along with your fixes and it should fix it @Risusama

**Comment by Risusama** [comment]
Oh, nevermind. It did fix it. Made a goddamn typo apparently xD

**Comment by Alex9914** [comment]
lol

**Comment by Risusama** [comment]
Okay, so the errors with `skillweb/cl_skillweb.lua` and  `vgui/pclassselect.lua` are actually caused by the client not getting information that the NetworkStrings are already added in serverside with `init.lua` via function GM:AddNetworkStrings, which is called by GM:Initialize..

**Comment by Risusama** [comment]
It would seem I had a corrupt zombiesurvival installation on my server from all the on-the-fly changes xD
I just tested on a fresh server and the gamemode seems to be working just fine with the current changes.

I'll squash these commits a bit.

**Comment by Risusama** [comment]
Nevermind, I fucked up the rebase twice now haha. Just merge like this if it's okay, the gamemode at least works now.

**Review by arthurgeron — CHANGES_REQUESTED** [review]
All network strings are added in `JetBoom/zombiesurvival/gamemodes/zombiesurvival/gamemode/init.lua`,  move all of them there and it should work just as well. It's not good for performance to keep calling them dynamically, it should be done once in init.

**Review by Risusama — COMMENTED** [review]
Done.
It would seem my test setup was totally broken because of a map voting addon (which I then fixed).

Most of the NetworkStrings were already in place in 'gamemode/init.lua', only a few were missing.

File: gamemodes/zombiesurvival/gamemode/init.lua [hunk 1]
```diff
@@ -1480,6 +1483,7 @@ end
 
 function GM:LastHuman(pl)
 	if not LASTHUMAN then
+		util.AddNetworkString("zs_lasthuman")
```
Comment (arthurgeron) [inline gamemodes/zombiesurvival/gamemode/init.lua:1486]: still here

File: gamemodes/zombiesurvival/gamemode/init.lua [hunk 2]
```diff
@@ -1928,6 +1931,7 @@ function GM:EndRound(winner)
 		end
 	end
 
+	util.AddNetworkString("zs_endround")
```
Comment (arthurgeron) [inline gamemodes/zombiesurvival/gamemode/init.lua:1934]: should be in init

File: gamemodes/zombiesurvival/gamemode/obj_player_extend_sv.lua [hunk 3]
```diff
@@ -1212,6 +1212,7 @@ function meta:DoHulls(classid, teamid)
 		end
 	end
 
+	util.AddNetworkString("zs_dohulls")
```
Comment (arthurgeron) [inline gamemodes/zombiesurvival/gamemode/obj_player_extend_sv.lua:1215]: init

File: gamemodes/zombiesurvival/gamemode/sh_weaponquality.lua [hunk 4]
```diff
@@ -135,6 +135,7 @@ function GM:CreateWeaponOfQuality(i, orig, quality, classname, branch)
 	-- TODO: Refactor me to not use the full weapon class once all self.BaseClass calls have been removed
 
 	local wept = weapons.Get(classname)
+	wept.ClassName = newclass
```
Comment (arthurgeron) [inline gamemodes/zombiesurvival/gamemode/sh_weaponquality.lua:138]: what's this for?

**Review by arthurgeron — CHANGES_REQUESTED** [review]
getting better, still some things left

File: gamemodes/zombiesurvival/gamemode/init.lua (hunk 1 above)
Comment (Risusama) [reply gamemodes/zombiesurvival/gamemode/init.lua]: Fixed.

File: gamemodes/zombiesurvival/gamemode/init.lua (hunk 2 above)
Comment (Risusama) [reply gamemodes/zombiesurvival/gamemode/init.lua]: Fixed.

File: gamemodes/zombiesurvival/gamemode/obj_player_extend_sv.lua (hunk 3 above)
Comment (Risusama) [reply gamemodes/zombiesurvival/gamemode/obj_player_extend_sv.lua]: Fixed.

File: gamemodes/zombiesurvival/gamemode/sh_weaponquality.lua (hunk 4 above)
Comment (Risusama) [reply gamemodes/zombiesurvival/gamemode/sh_weaponquality.lua]: It seemed to fix the grey model and weapon quality issue, but after much more testing it also seems to have broken gunturret. You can still buy it, it'll be in your inventory, but you cannot deploy it. When you select the gunturret, it just flashed green and then the deploy thingy disappeared.

I removed it now. After that the grey models issue is back, along with the weapon qualities error:
[error.log](https://github.com/JetBoom/zombiesurvival/files/10694897/error.log)

**Comment by Risusama** [comment]
I give up. I have no clue on how to fix the sh_weaponquality.lua.
I tracked the breaking change to be https://github.com/Facepunch/garrysmod/commit/71a181f9220ec19f63aaa7da00a318a9b11c0b6b.

The gamemode should be usable if someone manages to fix that. It isn't going to be me, I'm too bad at this and I got a fever yesterday.

My apologies that I have to leave the PR like this, I hope someone has the time, will and strength to do the last stuff.

**Comment by arthurgeron** [comment]
I’ll take a look at this this weekend

**Comment by Risusama** [comment]
Closing PR as the problem is fixed in master.

**Pull request closed without merging**